src/io/iohk/scalanet/peergroup/TLSPeerGroup.scala

```diff
@@ -0,0 +1,298 @@
+package io.iohk.scalanet.peergroup
+
+import java.net.InetSocketAddress
+import java.security.PrivateKey
+import java.security.cert.X509Certificate
+
+import io.iohk.decco.BufferInstantiator.global.HeapByteBuffer
+import io.iohk.decco._
+import io.iohk.scalanet.peergroup.PeerGroup.TerminalPeerGroup
+import io.iohk.scalanet.peergroup.TLSPeerGroup._
+import io.netty.bootstrap.{Bootstrap, ServerBootstrap}
+import io.netty.buffer.{ByteBuf, Unpooled}
+import io.netty.channel._
+import io.netty.channel.nio.NioEventLoopGroup
+import io.netty.channel.socket.SocketChannel
+import io.netty.channel.socket.nio.{NioServerSocketChannel, NioSocketChannel}
+import io.netty.handler.codec.bytes.ByteArrayEncoder
+import io.netty.handler.codec.{LengthFieldBasedFrameDecoder, LengthFieldPrepender}
+import io.netty.handler.ssl.{SslContextBuilder, SslHandshakeCompletionEvent}
+import io.netty.util
+import monix.eval.Task
+import monix.reactive.Observable
+import monix.reactive.subjects.{PublishSubject, ReplaySubject, Subject}
+import org.slf4j.LoggerFactory
+
+import scala.concurrent.Promise
+import scala.util.Success
+
+/**
+  * PeerGroup implementation on top of TCP.
+  * FIXME currently this class makes use of netty's LengthFieldPrepender to perform packet reassembly. This means
+  * the encoded bytes provided by the callers codec are not identical to the bytes put on the wire (since a
+  * length field is prepended to the byte stream). This class therefore cannot be used to talk to general services
+  * that are not instances of TCPPeerGroup.
+  *
+  * @param config bind address etc. See the companion object.
+  * @param codec a decco codec for reading writing messages to NIO ByteBuffer.
+  * @tparam M the message type.
+  */
+class TLSPeerGroup[M](val config: Config)(implicit codec: Codec[M]) extends TerminalPeerGroup[InetMultiAddress, M]() {
+
+  private val log = LoggerFactory.getLogger(getClass)
+
+  private[scalanet] class ServerChannelImpl[M](val nettyChannel: SocketChannel)(
+      implicit codec: Codec[M]
+  ) extends Channel[InetMultiAddress, M] {
+
+    private val log = LoggerFactory.getLogger(getClass)
+    private val messageSubject = ReplaySubject[M]()
+
+    private val sslServerCtx = SslContextBuilder.forServer(config.certChainPrivateKey, config.certChain: _*).build()
+
+    log.debug(
+      s"Creating server channel from ${nettyChannel.localAddress()} to ${nettyChannel.remoteAddress()} with channel id ${nettyChannel.id}"
+    )
+
+    nettyChannel
+      .pipeline()
+      .addLast("ssl", sslServerCtx.newHandler(nettyChannel.alloc())) //This needs to first
```

Comment:
It would be probably good idea to enforce `TLS 1.2` for both client and server 
```
val handler = sslContext.newHandler(socketChannel.alloc());
handler.engine().setEnabledProtocols(new String[] {"TLSv1.2"});
```
Another option worth considering, is to configure cipher suites which maintain perfect forward secrecy, in TLS those are cipher suites based on DH and ECDHE key exchanges.

It is worth noting that form TLS 1.3 only such cipher suites will be available, but unfortunatly TLS 1.3 is only available for java 11 which do not play good with scala yet.

Reply:
JDK 8 will use TLS 1.2 as default

Reply:
Agreed re ECDHE ciphers